### DISCUSSION TITLE
Add Output Schema Validation to next-safe-action

### DISCUSSION THREAD

**Discussion by ReemX** [next-safe-action/next-safe-action#280] [Ideas]
next-safe-action currently provides robust input validation for server actions using various schema validation libraries. To enhance type safety and consistency, I propose extending this functionality to include output validation as well.

## Feature Description

Implement a mechanism to define and validate the output schema of server actions, similar to how input schemas are currently handled. This would allow developers to:

1. Ensure consistency in action responses
2. Catch potential runtime errors early in the development process
3. Improve TypeScript inference for action results

## Proposed API

Extend the current API to include an `outputSchema` option:

```typescript
const myAction = actionClient
  .schema(inputSchema)
  .outputSchema(outputSchema)
  .action(async ({ parsedInput }) => {
    // Action implementation
    return result; // This would be validated against outputSchema
  });
```

## Benefits

- **Increased Type Safety**: Ensure that server actions always return data in the expected format.
- **Improved Developer Experience**: Catch inconsistencies between expected and actual output early in the development process.
- **Better Documentation**: Output schemas serve as self-documenting code for action results.
- **Enhanced Refactoring**: Changes to action output structures are caught at compile-time, facilitating safer refactoring.

## Implementation Considerations

- Support the same validation libraries as input schemas (Zod, Valibot, Yup, TypeBox)
- Provide clear error messages when output validation fails
- Consider performance implications of additional runtime checks

## Discussion Points

- Should output validation be optional or required?
- How should validation errors be handled (e.g., throw an error, return a specific error object)?
- Are there any potential drawbacks or edge cases that should be considered?

I believe this feature would significantly enhance the already powerful next-safe-action library, providing developers with even more tools to create robust and type-safe server actions.

**Comment by TheEdoRan** [comment]
Hi, this feature is already implemented in the library: [`outputSchema` docs](https://next-safe-action.dev/docs/define-actions/instance-methods#outputschema).

**Reply by ReemX** [reply]
Yep just noticed now :P
